File: ios/Shared/WidgetSnapshot.swift
import Foundation

/// Cross-target snapshot the main app publishes + the widget extension
/// reads (US-190). Widgets run in their own process and can't touch the
/// app's SwiftData store, so the main app computes a tiny rollup after
/// each sync and writes it to the shared App Group container as JSON.
/// The widget's TimelineProvider reads it back — no DB access, no
/// network, instant render.
///
/// Both targets compile this file (project.yml lists Shared/ under both
/// sources). Reuses the same App Group as ``IntakeInbox``.
public struct WidgetSnapshot: Codable, Equatable, Sendable {
    /// When the main app computed this rollup. The widget shows a
    /// relative "Updated 5m ago" so a stale snapshot is obvious rather
    /// than silently wrong.
    public let generatedAt: Date

    /// False when no one's signed in — the widget swaps to a
    /// "Sign in to see your stats" prompt instead of showing zeros that
    /// look like a dead business.
    public let isSignedIn: Bool

    /// Listings currently in the `active` state across all platforms.
    public let activeListings: Int

    /// Sales whose saleDate falls on the current local day.
    public let soldTodayCount: Int
    /// Gross sale price summed across today's sales (before fees).
    public let soldTodayGross: Double

    /// Sales that have sold but don't yet carry a payout reference —
    /// money earned but not yet in the bank.
    public let pendingPayoutCount: Int
    /// Net of those pending sales: sale price minus platform fees.
    public let pendingPayoutNet: Double

    /// US-1161: the user's currency override (ISO 4217), or nil to follow the
    /// device locale. Plumbed through so the widget + Siri summary format money
    /// in the same currency as the in-app surfaces instead of forcing USD.
    /// Optional so snapshots written before this field still decode.
    public let currencyCode: String?

    public init(
        generatedAt: Date,
        isSignedIn: Bool,
        activeListings: Int,
        soldTodayCount: Int,
        soldTodayGross: Double,
        pendingPayoutCount: Int,
        pendingPayoutNet: Double,
        currencyCode: String? = nil
    ) {
        self.generatedAt = generatedAt
        self.isSignedIn = isSignedIn
        self.activeListings = activeListings
        self.soldTodayCount = soldTodayCount
        self.soldTodayGross = soldTodayGross
        self.pendingPayoutCount = pendingPayoutCount
        self.pendingPayoutNet = pendingPayoutNet
        self.currencyCode = currencyCode
    }

    /// True when every rollup figure matches `other`, ignoring `generatedAt`.
    /// Used by the publisher (US-637) to skip a widget-timeline reload when the
    /// numbers haven't actually changed since the last publish.
    public func hasSameRollup(as other: WidgetSnapshot) -> Bool {
        isSignedIn == other.isSignedIn
            && activeListings == other.activeListings
            && soldTodayCount == other.soldTodayCount
            && soldTodayGross == other.soldTodayGross
            && pendingPayoutCount == other.pendingPayoutCount
            && pendingPayoutNet == other.pendingPayoutNet
            && currencyCode == other.currencyCode
    }

    /// Signed-out placeholder. Distinct from an all-zero signed-in
    /// snapshot (a real seller with no sales today) so the widget can
    /// branch its copy.
    public static func signedOut(generatedAt: Date = .now) -> WidgetSnapshot {
        WidgetSnapshot(
            generatedAt: generatedAt,
            isSignedIn: false,
            activeListings: 0,
            soldTodayCount: 0,
            soldTodayGross: 0,
            pendingPayoutCount: 0,
            pendingPayoutNet: 0
        )
    }

    /// Placeholder for the widget gallery + the brief window before the
    /// app has published its first real snapshot.
    public static let placeholder = WidgetSnapshot(
        generatedAt: .now,
        isSignedIn: true,
        activeListings: 12,
        soldTodayCount: 3,
        soldTodayGross: 184.0,
        pendingPayoutCount: 5,
        pendingPayoutNet: 312.55
    )
}

/// File-based store for the snapshot in the App Group container. Mirrors
/// ``IntakeInbox``'s container conventions so both features share one
/// group entitlement.
public enum WidgetSnapshotStore {
    public static let appGroupIdentifier = "group.com.gradethread.app"
    private static let filename = "widget-snapshot.json"

    public static var containerURL: URL? {
        FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: appGroupIdentifier)
    }

    public static var fileURL: URL? {
        containerURL?.appendingPathComponent(filename)
    }

    /// Writes the snapshot atomically. No-op (returns false) when the
    /// App Group container isn't available — e.g. unit tests without the
    /// entitlement; callers don't treat that as an error.
    @discardableResult
    public static func write(_ snapshot: WidgetSnapshot) -> Bool {
        guard let url = fileURL else { return false }
        do {
            let data = try encoder.encode(snapshot)
            // US-658: encrypt the cross-process snapshot (sales/payout figures)
            // at rest; the widget reads it after first unlock.
            try data.write(to: url, options: [.atomic, .completeFileProtectionUntilFirstUserAuthentication])
            return true
        } catch {
            return false
        }
    }

    /// Reads the last published snapshot, or nil when nothing's been
    /// written yet / the file is unreadable.
    public static func read() -> WidgetSnapshot? {
        guard let url = fileURL,
              let data = try? Data(contentsOf: url) else { return nil }
        return try? decoder.decode(WidgetSnapshot.self, from: data)
    }

    private static let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.dateEncodingStrategy = .iso8601
        return e
    }()

    private static let decoder: JSONDecoder = {
        let d = JSONDecoder()
        d.dateDecodingStrategy = .iso8601
        return d
    }()
}

File: android/app/src/main/java/com/gradethread/app/widget/WidgetSnapshot.kt
package com.gradethread.app.widget

import kotlinx.serialization.Serializable

/**
 * US-1380 (iOS `WidgetSnapshot`): the tiny rollup the home-screen widget draws.
 *
 * A widget render has a hard budget and no business opening Room or the
 * network, so the app computes this after every sync and writes it once.
 * The widget only ever reads it back.
 *
 * Every field is defaulted because this is a PERSISTED shape: a snapshot
 * written by an older build has to keep decoding after an upgrade, and a
 * missing field must not throw inside a render nobody can see fail.
 */
@Serializable
data class WidgetSnapshot(
    /**
     * When the app computed this. The widget shows "Updated 5m ago", so a
     * stale snapshot looks stale rather than looking wrong.
     */
    val generatedAt: Long = 0L,

    /**
     * False when nobody is signed in. The widget then asks them to sign in
     * rather than showing zeros, which read as a dead business.
     */
    val isSignedIn: Boolean = false,

    /** Live listings across every platform (active + relisted). */
    val activeListings: Int = 0,

    /** Completed sales whose sale date falls on the current LOCAL day. */
    val soldTodayCount: Int = 0,

    /** Gross of those sales, before fees. */
    val soldTodayGross: Double = 0.0,

    /** Completed sales with no payout reference — earned, not yet banked. */
    val pendingPayoutCount: Int = 0,

    /** Net of those: sale price minus platform fees, never below zero. */
    val pendingPayoutNet: Double = 0.0,
) {
    /**
     * Whether the NUMBERS match, ignoring [generatedAt].
     *
     * The publisher diffs on this so a sync that changed nothing doesn't
     * rewrite the file and wake every widget on the home screen.
     */
    fun hasSameRollup(other: WidgetSnapshot): Boolean =
        isSignedIn == other.isSignedIn &&
            activeListings == other.activeListings &&
            soldTodayCount == other.soldTodayCount &&
            soldTodayGross == other.soldTodayGross &&
            pendingPayoutCount == other.pendingPayoutCount &&
            pendingPayoutNet == other.pendingPayoutNet

    companion object {
        /**
         * Signed out.
         *
         * Deliberately distinct from an all-zero SIGNED-IN snapshot, which is a
         * real seller having a quiet day. The widget branches its copy on it.
         */
        fun signedOut(generatedAt: Long = 0L) = WidgetSnapshot(
            generatedAt = generatedAt,
            isSignedIn = false,
        )

        /**
         * What the widget picker previews, and what shows in the gap before the
         * first real publish. Plausible numbers, so the preview reads as the
         * feature rather than as an error state.
         */
        val PLACEHOLDER = WidgetSnapshot(
            generatedAt = 0L,
            isSignedIn = true,
            activeListings = 12,
            soldTodayCount = 3,
            soldTodayGross = 184.0,
            pendingPayoutCount = 5,
            pendingPayoutNet = 312.55,
        )
    }
}
